device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/SasTokenProvider.java

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.sdk.iot.device;
+
+/**
+ * Interface for allowing users to control SAS token generation. To see an example of how SAS tokens can be generated
+ * from device connection strings, see {@link com.microsoft.azure.sdk.iot.device.auth.IotHubSasToken}.
+ * @see <a href="https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-security#example">This document</a> for more details on sas tokens.
+ */
+public interface SasTokenProvider
+{
+    /**
+     * Returns a SAS token to be used for authentication.
+     * <p>
+     * This function will be called each time the client library needs a SAS token. This will follow different patterns
+     * based on which protocol your client object is using.
+     * <p>
+     * For HTTPS, this function will be called for each HTTPS request made (for instance, once per telemetry send), and does not need
+     * to be a unique token each time. Because of that, users may simply cache and return the same SAS token repeatedly here.
+     * Alternatively, users can generate very short-lived SAS tokens since they will only need to be valid for a relative short period of time.
+     * The user of this API is responsible for tracking when to renew the SAS token based on how long the previous token was valid for.
+     * <p>
+     * For AMQPS/AMQPS_WS, this function will be called once when first opening the connection, and then will be called again
```

Comment:
Is AMQPS a thing? or should this be AMQP?

Reply:
"AMQP" is to "AMQPS" what "HTTP" is to "HTTPS". Typically we say AMQPS because we do SSL handshakes when opening AMQP connections to secure them

Reply:
This doesn't apply to MQTT then?

Reply:
You'd think it would, but I don't think "MQTTS" is a thing, terminology-wise. Our MQTT connections always do an SSL handshake though, just like HTTPS and AMQPS/AMQPS_WS.